### DISCUSSION TITLE
How do I remove padding between fixed sized sprites in a sheet. And can I get the coordinates of each sprite once done.

### DISCUSSION THREAD

**Discussion by wyattbiker** [ImageMagick/ImageMagick#7051] [Help] [answered]
### ImageMagick version

any version

### Operating system, version and so on

Linux Kubuntu KDE

### Description

I have several sprite sheets and each sheet has sprites of fixed width/height. I would like to tell the tool the size of the sprites and have it remove all the empty space above/sides, but also export the final result coordinates (e.g. json or csv)? 

Maybe a 2 or 3 step process, (ie, extract the images, clean and then put back together)

For example of a sprites sheet, see below. This is a 600x100 sheet with with 6 sprites. So each sprite is in a 100x100 frame.
I would like to remove all the space in between to a new sprites sheet and get the coordinates of each frame.

from this:
![ATTACK](https://github.com/ImageMagick/ImageMagick/assets/930478/177c0eb2-7bc9-4086-90a4-71456d2b2b6e)

to this:
![packed](https://github.com/ImageMagick/ImageMagick/assets/930478/103fdc50-2dc5-436e-9d66-32e58aa7082a)

**Comment by snibgo** [comment]
One method is (bash syntax):
```
magick.exe \
  sixsprites.png \
  -alpha extract \
  -scale x1! \
  -fill White +opaque Black \
  -colorspace Gray \
  -define "ftxt:format=\c\n" \
  ftxt: >x.txt
```

The text output has 600 text lines, one for each column of the image. Each line contains either "0" meaning there is no sprite in that column, or "65535" if there is a sprite in that column. (If your IM Q-number isn't Q16, the large number will be different.)

A script can read that text output, and do whatever you want. It could note where each sprite starts and ends, and crop the image accordingly.

**Comment by fmw42** [comment, marked as answer]
Here is a solution for you using connected components to get the bounding boxes of the sprites from the alpha channel writing to a multi-page MIFF file via a sub-shell process. Then take all the bounding boxes and crop the image and smush (append) them with desired spacing between them.

Input:

![sprite](https://github.com/ImageMagick/ImageMagick/assets/13458218/2b8447b2-48fc-49ea-868b-121738eb0afb)

Unix Syntax:

```
spacing=10    # desired spacing
bboxArr=(`magick sprite.png \
-alpha extract -type bilevel \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:sort=X \
-connected-components 8 null: | grep "gray(255)" | awk '{print $2}'`)
num=${#bboxArr[*]}
( 
for ((i=0; i<num; i++)); do
bbox="${bboxArr[$i]}"
magick sprite.png -crop $bbox +repage miff:-
done \
) | magick - -background none +smush $spacing sprite_compressed.png

```

![sprite_compressed](https://github.com/ImageMagick/ImageMagick/assets/13458218/578aede0-e5af-4405-bdcc-ccf732cde30b)

**Reply by wyattbiker** [reply]
That is what I am looking for. Thank you.. . For anyone else looking to do this, I  aligned to ground by adding **-gravity south** for the sprites before the +smush.
